Lib/tkinter/__init__.py

```diff
@@ -498,10 +498,14 @@ def trace_variable(self, mode, callback):
 
         Return the name of the callback.
 
-        This deprecated method wraps a deprecated Tcl method that will
-        likely be removed in the future.  Use trace_add() instead.
+        This deprecated method wraps a deprecated Tcl method removed
+        in Tcl 9.0.  Use trace_add() instead.
         """
-        # TODO: Add deprecation warning
+        import warnings
+        warnings.warn(
+                "trace_variable() is deprecated and not supported with Tcl 9, "
+                "use trace_add() instead",
+                DeprecationWarning, stacklevel=2)
         cbname = self._register(callback)
         self._tk.call("trace", "variable", self._name, mode, cbname)
```

Comment:
Should this line be wrapped in `try...except  TclError...` for when run with tcl/tk 9?

Reply:
And what to do after catching the exception?

Reply:
Raise a better error message -- IF the TclError message is unclear, such as not specifying that failure us 9.0 specific.  I don't have 9.0 installed to test this.

Reply:
The error message is "`bad option "variable": must be add, info, or remove`". It may be not so clear from Python's point of view, but you get a deprecation warning emitted immediately before error. I think that it is not worth to change the error. The error can also be raised for other reasons, so we would need to rely on parsing the error message, which can be changed in future.

Reply:
I was thinking about the case when people have DeprecationWarnings turned off.  A possibility would be, when running tk9+, to immediately raise an error with more direct message 'This method does not exist in tk9, use...' right in trace_xyz itself.

Reply:
Do you suggest to write something like this?
```py
        try:
            self._tk.call("trace", "variable", self._name, mode, cbname)
        except _tkinter.TclError as err:
            if str(err) == 'bad option "variable": must be add, info, or remove':
                raise TypeError('trace_variable() is not supported '
                                'with Tcl 9; use trace_add() instead')
            raise
```
This is too verbose and fragile. If they change the error message it will not work.